rpc/wrtc_peer.go

```diff
@@ -244,12 +251,84 @@ func newPeerConnectionForServer(
 // DataChannel is open and available for renegotiation. The second Go channel will close when the
 // negotiation DataChannel is closed. PeerConnection.Close does not wait on DataChannel's to finish
 // their work. Thus waiting on this can be helpful to guarantee background goroutines have exitted.
-func ConfigureForRenegotiation(peerConn *webrtc.PeerConnection, logger golog.Logger) (<-chan struct{}, <-chan struct{}, error) {
+func ConfigureForRenegotiation(
+	peerConn *webrtc.PeerConnection,
+	role PeerRole,
+	logger golog.Logger,
+) (<-chan struct{}, <-chan struct{}, error) {
 	var negMu sync.Mutex
 
 	// All of Viam's PeerConnections hard code the `data` channel to be ID 0 and the `negotiation`
 	// channel to be ID 1. Thus these channels are "pre-negotiated".
 	negotiated := true
+
+	// The pion webrtc library may invoke `OnNegotiationNeeded` prior to the connection being
+	// established. We drop those requests on the floor. The original connection is established with
+	// our signaling and answering machinery.
+	//
+	// Additionally, just because a PeerConnection has moved into the `connected` state, that does
+	// not imply the pre-negotiated `negotiation` DataChannel is available for use. We return this
+	// `negOpened` channel to let tests create a happens-before relationship. Such that these tests
+	// can know when a PeerConnection method that invokes `OnNegotiationNeeded` can utilize this
+	// negotiation channel.
+	negOpened := make(chan struct{})
+
+	// negotiationChannel being set to a non-nil value is synchronized *after* negOpened is closed.
+	var negotiationChannel *webrtc.DataChannel
+
+	// OnNegotiationNeeded is webrtc callback for when a PeerConnection is mutated in a way such
+	// that its local description should change. Such as when a video track is added that should be
+	// streamed to the peer.
+	//
+	// Dan: The existing `OnNegotiationNeeded` algorithm is suitable when one side initiates all of
+	// the renegotiations. But it is not obvious that algorithm is suitable for when both sides can
+	// race on renegotiating. For now we "uninstall" the `OnNegotiationNeeded` callback and only
+	// allow the "server" to start a renegotiation.
+	if role == PeerRoleServer {
```

Comment:
I'm not observing where we set up an `OnNegotiationNeeded` handler for clients. I was expecting to still see the empty one. My understanding was that it was necessary to install a handler to ensure the renegotiation loop ends up in the right state. Was that intentional?

Reply:
The empty one doesn't do anything so yeah, it's not needed. All setting it does is ultimately have it be called if negotiation is needed. I think we'd have seen a test fail here if it were needed.

Reply:
Oh, is it not needed anymore because of [this](https://github.com/pion/webrtc/commit/b3856ffcfcd0eb8ae610bc72d407433ed86b39e4) patch? Before we had this [defer block](https://github.com/pion/webrtc/blob/952fbe068c0fc24dfde2c0788899827747ff12f0/peerconnection.go#L326-L334) and its placement relative to the "nil on-negotiation needed" check was important for preventing the problem.

Reply:
Even that I don't think matters. The client is never the one cause negotiation needed, so that part of the state machine ends up not mattering. The state for negotiation, like being stable, happens in set description.

Reply:
But if ever in the future the client would say add a track or a data channel on demand from its side, it would need the handler, otherwise it would miss negotiation needed events and not ever be stable.